--- a/lbuffer.c
+++ b/lbuffer.c
@@ -243,12 +243,8 @@ static uint32_t read_int32(const char *s, int bigendian, int wide) {
         }
     }
     else {
-        switch (wide) {
-        default: return 0;
-        case 4: n |= (*s++ & 0xFF) << 24;
-        case 3: n |= (*s++ & 0xFF) << 16;
-        case 2: n |= (*s++ & 0xFF) << 8;
-        case 1: n |= (*s++ & 0xFF);
-        }
+        int i;
+        for (i = 0; i < wide; ++i)
+            n |= (uint32_t)((unsigned char)s[i]) << (i * 8);
     }
     return n;
 }
